### PR TITLE
Add Azure support

### PR DESCRIPTION
**Pull request by justaugustus**
Adds support for instantiating Typhoon clusters on Microsoft Azure.

### CODE REVIEW

**Comment by justaugustus** [comment]
@dghubble functionally, everything works, but I still need to update docs for this.
I figured I'd toss it up here so you could take a peep!

File: azure/container-linux/kubernetes/controllers.tf [hunk 1]
```diff
@@ -0,0 +1,330 @@
+# TODO: Add Scale Sets implementation once support exists: https://github.com/kubernetes/kubernetes/issues/43287
+
+# Discrete DNS records for each controller's private IPv4 for etcd usage
+resource "azurerm_dns_a_record" "etcds" {
```
Comment (colemickens) [inline azure/container-linux/kubernetes/controllers.tf]: Why are separate etcd records needed? (aka, why not just use static IPs?)

File: docs/azure.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
+# AWS
```
Comment (colemickens) [inline docs/azure.md]: `s/AWS/Azure/g`

File: docs/azure.md [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+Reference the DNS zone id with `"${aws_route53_zone.zone-for-clusters.zone_id}"`.
+
+!!! tip ""
+    If you have an existing domain name with a zone file elsewhere, just carve out a subdomain that can be managed on Route53 (e.g. aws.mydomain.com) and [update nameservers](http://docs.aws.amazon.com/Route53/latest/DeveloperGuide/SOA-NSrecords.html).
+
+### Optional
+
+| Name | Description | Default | Example |
+|:-----|:------------|:--------|:--------|
+| controller_count | Number of controllers (i.e. masters) | 1 | 1 |
```
Comment (colemickens) [inline docs/azure.md]: More Amazon stuff in here, plus I think you require an extra thing or two for Azure.

Comment (colemickens) [reply docs/azure.md]: Actually there's quite a bit above two in the examples, even in places other than example outputs.

**Comment by colemickens** [comment]
Looks like it requires a full DNS zone to donate to the cluster? Or it just has to be present so that this can poke records into it? (Is it safe to assume the records have the cluster_name in it, or something to preserve uniqueness?)

File: docs/azure.md (hunk 2 above)
Comment (justaugustus) [reply docs/azure.md]: @colemickens Ya, still working the docs. This is just copy pasta for now.

**Comment by dghubble** [comment]
This gets the same answer [VMWare](https://github.com/poseidon/typhoon/issues/94) and OVH did.

Writing a module isn't the issue, its the maintenance, validation, testing, and support that goes into each platform, for Typhoon releases, for the foreseeable future. Typhoon clusters are run continuously on each platform, with workloads (simmer tests, not just "does it launch"). Platforms get independently tested and certified for conformance. Addons are verified to play nicely and provide a good admin experience. 

A priority ranked roadmap does include new platforms (further out), with QEMU/KVM support as the next platform planned, after other priorities. Azure is a good contender for another platform in my mind, but at a minimum we'd need to get Azure non-profit sponsorship to fund clusters. And preferably get a solid, trusted point of contact there.

We can revisit this in the future.

File: azure/container-linux/kubernetes/controllers.tf (hunk 1 above)
Comment (justaugustus) [reply azure/container-linux/kubernetes/controllers.tf]: @colemickens These etcd records are used as SANs in the etcd TLS asset generation process.

**Comment by justaugustus** [comment]
@colemickens I've updated the docs for Azure, if you'd like to take a look.
w.r.t. DNS records, you can delegate a domain or subdomain for usage in Azure. 
We'd simply be putting records into a pre-delegated zone.

While `cluster_name` is used to generate the records, it's not truly possible to ensure uniqueness if someone decides to create another cluster in that zone with the same name.

**Comment by justaugustus** [comment]
(for anyone driving by this PR...)
@dghubble and I have discussed this offline and there's some legwork to do before this might be considered for use in Typhoon, so I'm going to close this out.

In the meantime, if you're itching to use Typhoon on Azure, you can check out [typhoon-azure](https://github.com/justaugustus/typhoon-azure)!

**Pull request closed without merging**